src/extension/wallet/send/choose-fee.tsx

```diff
@@ -0,0 +1,342 @@
+      history.push({
+        pathname: SEND_CONFIRMATION_ROUTE,
+      });
+    } catch (error: any) {
+      handleError(error);
+    } finally {
+      setLoading(false);
+    }
+  };
+
+  const handleBackBtn = () => {
+    history.push(SEND_ADDRESS_AMOUNT_ROUTE);
+  };
+
+  const handleError = (err: unknown) => {
+    console.error(err);
+    setError(extractErrorMessage(err));
+    setUnsignedPset(undefined);
+    setFeeStr(undefined);
+    setSelectedFeeAsset(undefined);
+  };
+
+  const chooseFeeAndCreatePset = async (assetHash: string) => {
+    if (selectedFeeAsset === assetHash || !network) {
+      return; // skip if the same asset is selected
+    }
+
+    if (!recipient) {
+      throw new Error('address/amount to send not found');
+    }
+
+    try {
+      setLoading(true);
+      setSelectedFeeAsset(assetHash);
+
+      const pset = Creator.newPset();
+      const coinSelection = await walletRepository.selectUtxos(
+        network,
+        [{ asset: recipient.asset, amount: recipient.amount }],
+        true,
+        MainAccount,
+        MainAccountLegacy,
+        MainAccountTest
+      );
+
+      const updater = new Updater(pset);
+
+      // get the witness utxos from repository
+      const witnessUtxos = await Promise.all(
+        coinSelection.utxos.map((utxo) => {
+          return walletRepository.getWitnessUtxo(utxo.txID, utxo.vout);
+        })
+      );
+
+      updater.addInputs(
+        coinSelection.utxos.map((utxo, i) => ({
+          txid: utxo.txID,
+          txIndex: utxo.vout,
+          sighashType: Transaction.SIGHASH_ALL,
+          witnessUtxo: witnessUtxos[i],
+        }))
+      );
+
+      updater.addOutputs([
+        {
+          asset: recipient.asset,
+          amount: recipient.amount,
+          script: address.toOutputScript(recipient.address, networks[network]),
+          blinderIndex: 0,
+          blindingPublicKey: address.fromConfidential(recipient.address).blindingKey,
+        },
+      ]);
+
+      if (coinSelection.changeOutputs && coinSelection.changeOutputs.length > 0) {
+        const changeAddress = await mainAccount?.getNextAddress(true);
+        if (!changeAddress) {
+          throw new Error('change address not found');
+        }
+
+        updater.addOutputs([
+          {
+            asset: coinSelection.changeOutputs[0].asset,
+            amount: coinSelection.changeOutputs[0].amount,
+            script: address.toOutputScript(changeAddress, networks[network]),
+            blinderIndex: 0,
+            blindingPublicKey: address.fromConfidential(changeAddress).blindingKey,
+          },
+        ]);
+      }
+
+      const feeRate = 0.001;
+
+      const feeAmount = Math.floor(updater.pset.unsignedTx().virtualSize() * feeRate);
+
+      if (recipient.asset === networks[network].assetHash && updater.pset.outputs.length > 1) {
+        // subtract fee from change output
+        updater.pset.outputs[1].value = updater.pset.outputs[1].value - feeAmount;
```

Comment:
Does it mean that we can substract fee from change like in Bitcoin OR adding a specific fee output?
Is it new to Pset v2 ?

Reply:
No we must have a fee output in every cases. this line update the change amount if it is an L-BTC one. It prevents reselecting utxos if we have a change covering the fees.

Reply:
In current code fee output is only added in the reselect `else` block